### PR TITLE
manifest: ability to parse manifest or west section in manifest file

### PR DESCRIPTION
**Pull request by tejlmand**
Signed-off-by: Torsten Rasmussen <torsten.rasmussen@nordicsemi.no>

### CODE REVIEW

File: src/west/commands/project.py [hunk 1]
```diff
@@ -830,31 +830,14 @@ def _special_project(args, name):
...
-    # here.
-    project.url = url
-
-    return project
-
+        return Project(name, None, None, revision=revision,
+                       path=os.path.join('west', name), url=url)
+    
+    projects = Manifest.from_file(_manifest_path(args), name).projects
+    return next(proj for proj in projects if proj.name is name)
```
Comment (ulfalizer) [inline src/west/commands/project.py]: Comparing two strings with `is` isn't safe in general:

```python
>>> 100*'x' is 100*'x'
False
```

Had to do a double take on the `next()` thing. I might've gone for a plain `for` loop here to make it a bit less cryptic, though I'm usually into shortening stuff. :P

File: src/west/commands/project.py [hunk 2]
```diff
@@ -830,31 +830,14 @@ def _special_project(args, name):
...
-                      path=os.path.join('west', name))
-
-    # This could also be the name of a Git remote. The naming breaks a bit
-    # here.
-    project.url = url
-
-    return project
-
+        return Project(name, None, None, revision=revision,
+                       path=os.path.join('west', name), url=url)
```
Comment (ulfalizer) [inline src/west/commands/project.py]: Subjective, but could fold `url` and `revision` into the returned value.

File: src/west/manifest.py [hunk 3]
```diff
@@ -102,18 +102,17 @@ def __init__(self, source_file=None, source_data=None, sections=META_NAMES):
...
-        if 'manifest' not in self._data:
+        if section in self._data:
+            try:
+                pykwalify.core.Core(
+                            source_data=self._data[section],
+                            schema_files=[_SCHEMA_PATH[section]]
+                        ).validate()
+            except pykwalify.errors.SchemaError as e:
+                        self._malformed(e, section)
+        elif section is not 'west':
```
Comment (ulfalizer) [inline src/west/manifest.py]: Unsafe string comparison with `is`.

File: src/west/manifest.py [hunk 4]
```diff
@@ -144,26 +143,29 @@ def _malformed(self, complaint, section='manifest'):
...
+        manifest = data.get(section)
+
+        if section is 'west':
+            if manifest is None:
+                manifest = {'url': WEST_URL_DEFAULT,
+                            'revision': WEST_REV_DEFAULT}
+
+            project = Project(section,
+                              None,
+                              None,
```
Comment (ulfalizer) [inline src/west/manifest.py]: Could name the arguments (`foo=None`) to make it clearer what the `None`s mean.

File: src/west/manifest.py [hunk 5]
```diff
@@ -144,26 +143,29 @@ def _malformed(self, complaint, section='manifest'):
...
+                              url=manifest['url'],
+                              revision=manifest['revision'],
+                              path=os.path.join('west', section))
+            projects.append(project)
+            self.projects = tuple(projects)
```
Comment (ulfalizer) [inline src/west/manifest.py]: Could maybe have something like this:

```python
# Note: This is a single-element tuple
self.projects = (Project(section, ...),)
```

These could then be moved later (to as close as possible to where they're used):

```python
project = []
project_abspaths = set()
```

Having a plain list for `self.projects` would be fine too, imo (including below).

File: src/west/manifest.py [hunk 6]
```diff
@@ -315,7 +297,7 @@ class Project:
     __slots__ = 'name remote url path abspath clone_depth revision'.split()
 
     def __init__(self, name, remote, defaults, path=None, clone_depth=None,
-                 revision=None):
+                 revision=None, url=None):
```
Comment (ulfalizer) [inline src/west/manifest.py]: Missing documentation for the new parameter in the docstring.

File: src/west/manifest.py [hunk 7]
```diff
@@ -102,18 +102,17 @@ def __init__(self, source_file=None, source_data=None, sections=META_NAMES):
...
 
-        if 'manifest' not in self._data:
+        if section in self._data:
+            try:
+                pykwalify.core.Core(
+                            source_data=self._data[section],
+                            schema_files=[_SCHEMA_PATH[section]]
+                        ).validate()
+            except pykwalify.errors.SchemaError as e:
+                        self._malformed(e, section)
```
Comment (ulfalizer) [inline src/west/manifest.py]: Over-indented line

**Review by ulfalizer — CHANGES_REQUESTED** [review]
Some random code nits

**Comment by carlescufi** [comment]
Options

```
self.special_projects = [ Project(west) ]
vs
self.west_project = Project(west)
vs
self.westmeta = WestMeta(west)
```

**Pull request closed without merging**